<a href="https://colab.research.google.com/github/nidhicodes4045/datascience442/blob/main/Project_Dataset_Cleaning1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Dataset

> #### Description

> The Food Access Research Atlas is a dataset that was developed by the United States Department of Agriculture's Economic Research Service (ERS) to understand and address the issue of limited access to healthy, affordable food across the US. This county-level dataset combines information from several sources, including the 2019 STARS directory of SNAP-authorized stores, a 2019 supermarket list, the 2010 Decennial Census, and the 2014-18 American Community Survey (ACS). By integrating data on store locations, demographics, and socioeconomic factors, the Atlas provides a comprehensive view of food access, enabling researchers to analyze the complex relationship between these factors and food insecurity within communities. Ryan Whitcomb, Joung Min Choi, and Bo Guan are associated with providing the dataset on CORGIS [Food Access](https://corgis-edu.github.io/corgis/csv/food_access/). It's possible they worked with the data after its initial creation by the ERS.

>The County Demographics dataset is a dataset originally compiled by the United States Census Bureau and made available through their QuickFacts tool.  It provides demographic information at the county level for the United States, designed to offer easy access to key population, business, and geographic statistics. This county-level dataset draws information from several Census Bureau sources, including the Decennial Census, the American Community Survey (ACS), the County Business Patterns (CBP), and the Population Estimates Program. By integrating data from these sources, the dataset provides a snapshot of key demographic indicators, enabling researchers to analyze various population characteristics. Ryan Whitcomb, Joung Min Choi, and Bo Guan are associated with providing the dataset on CORGIS [County Demographics](https://corgis-edu.github.io/corgis/csv/county_demographics/). It's possible they compiled or curated the data after its initial collection by the U.S. Census Bureau.

We have prepared a dataset that consists of a subset of these two CORGIS datasets. We will be referencing the data that we have prepared in our Github repository: ["food_df"]("https://raw.githubusercontent.com/nidhicodes4045/datascience442/refs/heads/main/Food_access_mod.csv") and ["county_df"]("https://raw.githubusercontent.com/nidhicodes4045/datascience442/refs/heads/main/county_demographics_mod.csv").

## Imports

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os
#!pip install us     #Just do it once
os.environ['DC_STATEHOOD'] = '1' # adds dc as a state
import seaborn as sns
import re

In [ ]:
import us

In [ ]:
print(os.getcwd())

/content


##Our Dataset initialization

In [ ]:
county_df = pd.read_csv("https://raw.githubusercontent.com/nidhicodes4045/datascience442/refs/heads/main/county_demographics_mod.csv")
food_df = pd.read_csv("https://raw.githubusercontent.com/nidhicodes4045/datascience442/refs/heads/main/Food_access_mod.csv")

##Dataframe info

###County df head and tail

In [ ]:
county_df.head()

County State  Age.Percent 65 and Older  \
0  Abbeville County    SC                      22.4   
1     Acadia Parish    LA                      15.8   
2   Accomack County    VA                      24.6   
3        Ada County    ID                      14.9   
4      Adair County    IA                      23.0   

  Age.Percent Under 18 Years Age.Percent Under 5 Years  \
0                       19.8                       4.7   
1                       25.8                       6.9   
2                       20.7                       5.6   
3                       23.2                       5.6   
4                       21.8                       5.6   

   Education.Bachelor's Degree or Higher  Education.High School or Higher  \
0                                   15.6                             81.7   
1                                   13.3                             79.0   
2                                   19.5                             81.5   
3                                   38.5                             95.2   
4                                   18.5                             94.2   

  Employment.Nonemployer Establishments  \
0                                  1416   
1                                  4533   
2                                  2387   
3                                 41464   
4                                   609   

  Ethnicities.American Indian and Alaska Native Alone Ethnicities.Asian Alone  \
0                                                0.3                      0.4   
1                                                0.4                      0.3   
2                                                0.7                      0.8   
3                                                0.8                      2.7   
4                                                0.3                      0.5   

   ... Population.Population per Square Mile  \
0  ...                                  51.8   
1  ...                                  94.3   
2  ...                                  73.8   
3  ...                                 372.8   
4  ...                                  13.5   

   Sales.Accommodation and Food Services Sales Sales.Retail Sales  \
0                                        12507              91371   
1                                        52706             602739   
2                                        53568             348195   
3                                       763099            5766679   
4                                                           63002   

  Employment.Firms.Total  Employment.Firms.Women-Owned  \
0                   1450                           543   
1                   4664                          1516   
2                   2997                           802   
3                  41789                         14661   
4                    914                           304   

   Employment.Firms.Men-Owned Employment.Firms.Minority-Owned  \
0                         689                             317   
1                        2629                             705   
2                        1716                             335   
3                       19409                            3099   
4                         499                               0   

   Employment.Firms.Nonminority-Owned Employment.Firms.Veteran-Owned  \
0                                1080                            187   
1                                3734                            388   
2                                2560                            212   
3                               36701                           3803   
4                                 861                            185   

  Employment.Firms.Nonveteran-Owned  
0                              1211  
1                              4007  
2                              2536  
3                             35132  
4                               679  

[5 rows x 43 colum

In [ ]:
county_df.tail()

County State  Age.Percent 65 and Older  \
3134     Yuma County    AZ                      19.3   
3135     Yuma County    CO                      18.7   
3136   Zapata County    TX                      13.2   
3137   Zavala County    TX                      14.6   
3138  Ziebach County    SD                       9.6   

     Age.Percent Under 18 Years Age.Percent Under 5 Years  \
3134                       25.1                       7.1   
3135                       27.4                       7.5   
3136                       33.1                       8.6   
3137                       28.4                       7.2   
3138                       27.5                       5.5   

      Education.Bachelor's Degree or Higher  Education.High School or Higher  \
3134                                   15.0                             73.3   
3135                                   21.8                             88.6   
3136                                   11.6                             61.9   
3137                                   10.9                             66.9   
3138                                   16.4                             84.1   

     Employment.Nonemployer Establishments  \
3134                                  9896   
3135                                  1020   
3136                                  1452   
3137                                   837   
3138                                    87   

     Ethnicities.American Indian and Alaska Native Alone  \
3134                                                2.3    
3135                                                  1    
3136                                                0.5    
3137                                                1.1    
3138                                        Unavailable    

     Ethnicities.Asian Alone  ... Population.Population per Square Mile  \
3134                     1.5  ...                                  35.5   
3135                     0.5  ...                                   4.2   
3136                     0.2  ...                                  14.0   
3137                     0.3  ...                                   9.0   
3138                     0.3  ...                                   1.4   

      Sales.Accommodation and Food Services Sales Sales.Retail Sales  \
3134                                       307540            1995974   
3135                                         8501             125565   
3136                                                           75681   
3137                                         8808              45596   
3138                                                           15757   

     Employment.Firms.Total  Employment.Firms.Women-Owned  \
3134                  10846                          4298   
3135                   1492                           391   
3136                   1964                           818   
3137                   1232                           486   
3138                     78                             0   

      Employment.Firms.Men-Owned Employment.Firms.Minority-Owned  \
3134                        4529                            5749   
3135                         797                              45   
3136                        1003                            1680   
3137                         674                            1062   
3138                          42                              29   

      Employment.Firms.Nonminority-Owned Employment.Firms.Veteran-Owned  \
3134                                4476                            839   
3135                                1350                             66   
3136                                 235                            181   
3137                                 159                             42   
3138                                  36                              0   

     Employment.Firms.Nonveteran-Owned  
3134                              9265 

###Food df head and tail

In [ ]:
food_df.head()

County  Population    State  \
0  Autauga County       54571  Alabama   
1  Baldwin County      182265  Alabama   
2  Barbour County       27457  Alabama   
3     Bibb County       22915  Alabama   
4   Blount County       57322  Alabama   

   Housing Data.Residing in Group Quarters  Housing Data.Total Housing Units  \
0                                      455                             20221   
1                                     2307                             73180   
2                                     3193                              9820   
3                                     2224                              7953   
4                                      489                             21578   

   Vehicle Access.1 Mile  Vehicle Access.1/2 Mile  Vehicle Access.10 Miles  \
0                    834                     1045                    222.0   
1                   1653                     2178                     32.0   
2                    545                      742                    201.0   
3                    312                      441                      0.0   
4                    752                      822                      0.0   

   Vehicle Access.20 Miles  Low Access Numbers.Children.1 Mile  ...  \
0                        0                                9973  ...   
1                        0                               30633  ...   
2                        0                                3701  ...   
3                        0                                4198  ...   
4                        0                               12575  ...   

   Low Access Numbers.Low Income People.10 Miles  \
0                                           2307   
1                                            846   
2                                           2440   
3                                            102   
4                                              0   

   Low Access Numbers.Low Income People.20 Miles  \
0                                              0   
1                                              0   
2                                              0   
3                                              0   
4                                              0   

   Low Access Numbers.People.1 Mile  Low Access Numbers.People.1/2 Mile  \
0                           37424.0                               49497   
1                          132442.0                              165616   
2                               NaN                               23762   
3                           17560.0                               20989   
4                           50848.0                               54933   

   Low Access Numbers.People.10 Miles  Low Access Numbers.People.20 Miles  \
0                                5119                                   0   
1                                2308                                   0   
2                                4643                                   0   
3                                 365                                   0   
4                                   0                                   0   

   Low Access Numbers.Seniors.1 Mile  Low Access Numbers.Seniors.1/2 Mile  \
0                               4393                               5935.0   
1                              21828                              27241.0   
2                               2537                               3348.0   
3                               2262                               2630.0   
4                               7114                               7810.0   

   Low Access Numbers.Seniors.10 Miles  Low Access Numbers.Seniors.20 Miles  
0                                  707                                    0  
1                                  390                                    0  
2                                  629                                    0  
3                                   72                            

In [ ]:
food_df.tail()

County  Population    State  \
3137  Sweetwater County       43806  Wyoming   
3138       Teton County       21294  Wyoming   
3139       Uinta County       21118  Wyoming   
3140    Washakie County        8533  Wyoming   
3141      Weston County        7208  Wyoming   

      Housing Data.Residing in Group Quarters  \
3137                                      679   
3138                                      271   
3139                                      270   
3140                                      140   
3141                                      313   

      Housing Data.Total Housing Units  Vehicle Access.1 Mile  \
3137                             16475                    284   
3138                              8973                     74   
3139                              7668                    175   
3140                              3492                     37   
3141                              3021                     22   

      Vehicle Access.1/2 Mile  Vehicle Access.10 Miles  \
3137                      372                     18.0   
3138                      144                      0.0   
3139                      283                      9.0   
3140                      106                      7.0   
3141                       43                      2.0   

      Vehicle Access.20 Miles  Low Access Numbers.Children.1 Mile  ...  \
3137                       13                                6280  ...   
3138                        0                                2853  ...   
3139                        0                                3755  ...   
3140                        6                                 746  ...   
3141                        0                                 735  ...   

      Low Access Numbers.Low Income People.10 Miles  \
3137                                            759   
3138                                            383   
3139                                            290   
3140                                            218   
3141                                            247   

      Low Access Numbers.Low Income People.20 Miles  \
3137                                            614   
3138                                            181   
3139                                              3   
3140                                            176   
3141                                             55   

      Low Access Numbers.People.1 Mile  Low Access Numbers.People.1/2 Mile  \
3137                           24036.0                               36045   
3138                           15298.0                               18354   
3139                           12250.0                               17998   
3140                            2869.0                                4961   
3141                            3659.0                                5237   

      Low Access Numbers.People.10 Miles  Low Access Numbers.People.20 Miles  \
3137                                2548                                2063   
3138                                1677                                 572   
3139                                 726                                  14   
3140                                 902                                 730   
3141                                 840                                 188   

      Low Access Numbers.Seniors.1 Mile  Low Access Numbers.Seniors.1/2 Mile  \
3137                               2358                               3232.0   
3138                               1867                               2009.0   
3139                               1078                                  NaN   
3140                                506                                  NaN   
3141                                518                                824.0   

      Low Access Numbers.Seniors.10 Miles  Low Access Numbers.Seniors.20 Miles  
3137                                  255                                 

###county df info

In [ ]:
county_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3139 entries, 0 to 3138
Data columns (total 43 columns):
 #   Column                                                        Non-Null Count  Dtype  
---  ------                                                        --------------  -----  
 0   County                                                        3139 non-null   object 
 1   State                                                         3139 non-null   object 
 2   Age.Percent 65 and Older                                      3139 non-null   float64
 3   Age.Percent Under 18 Years                                    3139 non-null   object 
 4   Age.Percent Under 5 Years                                     3139 non-null   object 
 5   Education.Bachelor's Degree or Higher                         3139 non-null   float64
 6   Education.High School or Higher                               3139 non-null   float64
 7   Employment.Nonemployer Establishments                         3139 no

###food df info

In [ ]:
food_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3142 entries, 0 to 3141
Data columns (total 25 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   County                                         3142 non-null   object 
 1   Population                                     3142 non-null   int64  
 2   State                                          3142 non-null   object 
 3   Housing Data.Residing in Group Quarters        3142 non-null   int64  
 4   Housing Data.Total Housing Units               3142 non-null   int64  
 5   Vehicle Access.1 Mile                          3142 non-null   int64  
 6   Vehicle Access.1/2 Mile                        3142 non-null   int64  
 7   Vehicle Access.10 Miles                        2807 non-null   float64
 8   Vehicle Access.20 Miles                        3142 non-null   int64  
 9   Low Access Numbers.Children.1 Mile             3142 

###null count

In [ ]:
print(food_df.isnull().sum())

County                                             0
Population                                         0
State                                              0
Housing Data.Residing in Group Quarters            0
Housing Data.Total Housing Units                   0
Vehicle Access.1 Mile                              0
Vehicle Access.1/2 Mile                            0
Vehicle Access.10 Miles                          335
Vehicle Access.20 Miles                            0
Low Access Numbers.Children.1 Mile                 0
Low Access Numbers.Children.1/2 Mile               0
Low Access Numbers.Children.10 Miles               0
Low Access Numbers.Children.20 Miles               0
Low Access Numbers.Low Income People.1 Mile        0
Low Access Numbers.Low Income People.1/2 Mile      0
Low Access Numbers.Low Income People.10 Miles      0
Low Access Numbers.Low Income People.20 Miles      0
Low Access Numbers.People.1 Mile                 156
Low Access Numbers.People.1/2 Mile            


#Data Cleaning Step 1: Rectify structural misalignments
1. Check shape of the data – number of columns and rows  
2. Make sure the columns are named for convenience, follow the data dictionary
3. Do rows need to be named? Would default 0 order index work?


In [ ]:
county_df_shape = county_df.shape
print(county_df_shape)

(3139, 43)


In [ ]:
food_df_shape = food_df.shape
print(food_df_shape)

(3142, 25)


Food df has 3 more rows
county df has more cols

In [ ]:
#function to lowercase all text
def lowerandstriptext(text):
  newtext = re.sub(r"[\s\t]+", "", text)
  return newtext.lower().strip()

In [ ]:
#function to convert full state names to abbr
def convertstate(text):
  if us.states.lookup(text) is None:
    return "None"
  else:
    state = us.states.lookup(text).abbr
    statenew = re.sub(r"[\s\t]+", "", state)
    return statenew.lower().strip()

In [ ]:
#Apply the lowercase and strip functions to the State and County cols county df
county_df["County"] = county_df["County"].apply(lowerandstriptext)
county_df["State"] = county_df["State"].apply(lowerandstriptext)
county_df.head()

County State  Age.Percent 65 and Older Age.Percent Under 18 Years  \
0  abbevillecounty    sc                      22.4                       19.8   
1     acadiaparish    la                      15.8                       25.8   
2   accomackcounty    va                      24.6                       20.7   
3        adacounty    id                      14.9                       23.2   
4      adaircounty    ia                      23.0                       21.8   

  Age.Percent Under 5 Years  Education.Bachelor's Degree or Higher  \
0                       4.7                                   15.6   
1                       6.9                                   13.3   
2                       5.6                                   19.5   
3                       5.6                                   38.5   
4                       5.6                                   18.5   

   Education.High School or Higher Employment.Nonemployer Establishments  \
0                             81.7                                  1416   
1                             79.0                                  4533   
2                             81.5                                  2387   
3                             95.2                                 41464   
4                             94.2                                   609   

  Ethnicities.American Indian and Alaska Native Alone Ethnicities.Asian Alone  \
0                                                0.3                      0.4   
1                                                0.4                      0.3   
2                                                0.7                      0.8   
3                                                0.8                      2.7   
4                                                0.3                      0.5   

   ... Population.Population per Square Mile  \
0  ...                                  51.8   
1  ...                                  94.3   
2  ...                                  73.8   
3  ...                                 372.8   
4  ...                                  13.5   

   Sales.Accommodation and Food Services Sales Sales.Retail Sales  \
0                                        12507              91371   
1                                        52706             602739   
2                                        53568             348195   
3                                       763099            5766679   
4                                                           63002   

  Employment.Firms.Total  Employment.Firms.Women-Owned  \
0                   1450                           543   
1                   4664                          1516   
2                   2997                           802   
3                  41789                         14661   
4                    914                           304   

   Employment.Firms.Men-Owned Employment.Firms.Minority-Owned  \
0                         689                             317   
1                        2629                             705   
2                        1716                             335   
3                       19409                            3099   
4                         499                               0   

   Employment.Firms.Nonminority-Owned Employment.Firms.Veteran-Owned  \
0                                1080                            187   
1                                3734                            388   
2                                2560                            212   
3                               36701                           3803   
4                                 861                            185   

  Employment.Firms.Nonveteran-Owned  
0                              1211  
1                              4007  
2                              2536  
3                             35132  
4                               679  

[5 rows x 43 columns]

In [ ]:
#Apply the preprocessing functions to food df
food_df["County"] = food_df["County"].apply(lowerandstriptext)
food_df["State"] = food_df["State"].apply(convertstate)
food_df.head()

County  Population State  Housing Data.Residing in Group Quarters  \
0  autaugacounty       54571    al                                      455   
1  baldwincounty      182265    al                                     2307   
2  barbourcounty       27457    al                                     3193   
3     bibbcounty       22915    al                                     2224   
4   blountcounty       57322    al                                      489   

   Housing Data.Total Housing Units  Vehicle Access.1 Mile  \
0                             20221                    834   
1                             73180                   1653   
2                              9820                    545   
3                              7953                    312   
4                             21578                    752   

   Vehicle Access.1/2 Mile  Vehicle Access.10 Miles  Vehicle Access.20 Miles  \
0                     1045                    222.0                        0   
1                     2178                     32.0                        0   
2                      742                    201.0                        0   
3                      441                      0.0                        0   
4                      822                      0.0                        0   

   Low Access Numbers.Children.1 Mile  ...  \
0                                9973  ...   
1                               30633  ...   
2                                3701  ...   
3                                4198  ...   
4                               12575  ...   

   Low Access Numbers.Low Income People.10 Miles  \
0                                           2307   
1                                            846   
2                                           2440   
3                                            102   
4                                              0   

   Low Access Numbers.Low Income People.20 Miles  \
0                                              0   
1                                              0   
2                                              0   
3                                              0   
4                                              0   

   Low Access Numbers.People.1 Mile  Low Access Numbers.People.1/2 Mile  \
0                           37424.0                               49497   
1                          132442.0                              165616   
2                               NaN                               23762   
3                           17560.0                               20989   
4                           50848.0                               54933   

   Low Access Numbers.People.10 Miles  Low Access Numbers.People.20 Miles  \
0                                5119                                   0   
1                                2308                                   0   
2                                4643                                   0   
3                                 365                                   0   
4                                   0                                   0   

   Low Access Numbers.Seniors.1 Mile  Low Access Numbers.Seniors.1/2 Mile  \
0                               4393                               5935.0   
1                              21828                              27241.0   
2                               2537                               3348.0   
3                               2262                               2630.0   
4                               7114                               7810.0   

   Low Access Numbers.Seniors.10 Miles  Low Access Numbers.Seniors.20 Miles  
0                                  707                                    0  
1                                  390                                    0  
2                                  629                                    0  
3                                   72                                    0  
4          

In [ ]:
#Verify if any 'Nones' exist before merging
check_state_food_df = (food_df['State'] == 'None').sum()
print("The number of 'None' in food_df: " + str(check_state_food_df))

The number of 'None' in food_df: 0


In [ ]:
#Insert merged columns at index 2 and 3 for county and food dfs respectively
county_df.insert(2, "mergedkeycol", county_df['County'] + county_df['State'])
food_df.insert(3, "mergedkeycol", food_df['County'] + food_df['State'])

In [ ]:
#Check county df
county_df.head()

County State       mergedkeycol  Age.Percent 65 and Older  \
0  abbevillecounty    sc  abbevillecountysc                      22.4   
1     acadiaparish    la     acadiaparishla                      15.8   
2   accomackcounty    va   accomackcountyva                      24.6   
3        adacounty    id        adacountyid                      14.9   
4      adaircounty    ia      adaircountyia                      23.0   

  Age.Percent Under 18 Years Age.Percent Under 5 Years  \
0                       19.8                       4.7   
1                       25.8                       6.9   
2                       20.7                       5.6   
3                       23.2                       5.6   
4                       21.8                       5.6   

   Education.Bachelor's Degree or Higher  Education.High School or Higher  \
0                                   15.6                             81.7   
1                                   13.3                             79.0   
2                                   19.5                             81.5   
3                                   38.5                             95.2   
4                                   18.5                             94.2   

  Employment.Nonemployer Establishments  \
0                                  1416   
1                                  4533   
2                                  2387   
3                                 41464   
4                                   609   

  Ethnicities.American Indian and Alaska Native Alone  ...  \
0                                                0.3   ...   
1                                                0.4   ...   
2                                                0.7   ...   
3                                                0.8   ...   
4                                                0.3   ...   

  Population.Population per Square Mile  \
0                                  51.8   
1                                  94.3   
2                                  73.8   
3                                 372.8   
4                                  13.5   

  Sales.Accommodation and Food Services Sales  Sales.Retail Sales  \
0                                       12507               91371   
1                                       52706              602739   
2                                       53568              348195   
3                                      763099             5766679   
4                                                           63002   

  Employment.Firms.Total Employment.Firms.Women-Owned  \
0                   1450                          543   
1                   4664                         1516   
2                   2997                          802   
3                  41789                        14661   
4                    914                          304   

   Employment.Firms.Men-Owned  Employment.Firms.Minority-Owned  \
0                         689                              317   
1                        2629                              705   
2                        1716                              335   
3                       19409                             3099   
4                         499                                0   

  Employment.Firms.Nonminority-Owned  Employment.Firms.Veteran-Owned  \
0                               1080                             187   
1                               3734                             388   
2                               2560                             212   
3                              36701                            3803   
4                                861                             185   

  Employment.Firms.Nonveteran-Owned  
0                              1211  
1                              4007  
2                              2536  
3                             35132  
4                               679  

[5 rows x 44 columns]

In [ ]:
county_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3139 entries, 0 to 3138
Data columns (total 44 columns):
 #   Column                                                        Non-Null Count  Dtype  
---  ------                                                        --------------  -----  
 0   County                                                        3139 non-null   object 
 1   State                                                         3139 non-null   object 
 2   mergedkeycol                                                  3139 non-null   object 
 3   Age.Percent 65 and Older                                      3139 non-null   float64
 4   Age.Percent Under 18 Years                                    3139 non-null   object 
 5   Age.Percent Under 5 Years                                     3139 non-null   object 
 6   Education.Bachelor's Degree or Higher                         3139 non-null   float64
 7   Education.High School or Higher                               3139 no

In [ ]:
#Check food df
food_df.head()

County  Population State     mergedkeycol  \
0  autaugacounty       54571    al  autaugacountyal   
1  baldwincounty      182265    al  baldwincountyal   
2  barbourcounty       27457    al  barbourcountyal   
3     bibbcounty       22915    al     bibbcountyal   
4   blountcounty       57322    al   blountcountyal   

   Housing Data.Residing in Group Quarters  Housing Data.Total Housing Units  \
0                                      455                             20221   
1                                     2307                             73180   
2                                     3193                              9820   
3                                     2224                              7953   
4                                      489                             21578   

   Vehicle Access.1 Mile  Vehicle Access.1/2 Mile  Vehicle Access.10 Miles  \
0                    834                     1045                    222.0   
1                   1653                     2178                     32.0   
2                    545                      742                    201.0   
3                    312                      441                      0.0   
4                    752                      822                      0.0   

   Vehicle Access.20 Miles  ...  \
0                        0  ...   
1                        0  ...   
2                        0  ...   
3                        0  ...   
4                        0  ...   

   Low Access Numbers.Low Income People.10 Miles  \
0                                           2307   
1                                            846   
2                                           2440   
3                                            102   
4                                              0   

   Low Access Numbers.Low Income People.20 Miles  \
0                                              0   
1                                              0   
2                                              0   
3                                              0   
4                                              0   

   Low Access Numbers.People.1 Mile  Low Access Numbers.People.1/2 Mile  \
0                           37424.0                               49497   
1                          132442.0                              165616   
2                               NaN                               23762   
3                           17560.0                               20989   
4                           50848.0                               54933   

   Low Access Numbers.People.10 Miles  Low Access Numbers.People.20 Miles  \
0                                5119                                   0   
1                                2308                                   0   
2                                4643                                   0   
3                                 365                                   0   
4                                   0                                   0   

   Low Access Numbers.Seniors.1 Mile  Low Access Numbers.Seniors.1/2 Mile  \
0                               4393                               5935.0   
1                              21828                              27241.0   
2                               2537                               3348.0   
3                               2262                               2630.0   
4                               7114                               7810.0   

   Low Access Numbers.Seniors.10 Miles  Low Access Numbers.Seniors.20 Miles  
0                                  707                                    0  
1                                  390                                    0  
2                                  629                                    0  
3                                   72                                    0  
4                                    0                                    0  

[5 rows x 26 columns]

In [ ]:
county_df.dtypes

County                                                           object
State                                                            object
mergedkeycol                                                     object
Age.Percent 65 and Older                                        float64
Age.Percent Under 18 Years                                       object
Age.Percent Under 5 Years                                        object
Education.Bachelor's Degree or Higher                           float64
Education.High School or Higher                                 float64
Employment.Nonemployer Establishments                            object
Ethnicities.American Indian and Alaska Native Alone              object
Ethnicities.Asian Alone                                          object
Ethnicities.Black Alone                                          object
Ethnicities.Hispanic or Latino                                  float64
Ethnicities.Native Hawaiian and Other Pacific Islander Alone     object
Ethnicities.Two or More Races                                    object
Ethnicities.White Alone                                         float64
Ethnicities.White Alone\t not Hispanic or Latino                float64
Housing.Homeownership Rate                                       object
Housing.Households                                              float64
Housing.Housing Units                                            object
Housing.Median Value of Owner-Occupied Units                     object
Housing.Persons per Household                                   float64
Income.Median Houseold Income                                     int64
Income.Per Capita Income                                          int64
Miscellaneous.Foreign Born                                      float64
Miscellaneous.Land Area                                         float64
Miscellaneous.Language Other than English at Home               float64
Miscellaneous.Living in Same House +1 Years                     float64
Miscellaneous.Manufacturers Shipments                            object
Miscellaneous.Mean Travel Time to Work                           object
Miscellaneous.Percent Female                                    float64
Miscellaneous.Veterans                                           object
Population.2020 Population                                        int64
Population.2010 Population                                        int64
Population.Population per Square Mile                           float64
Sales.Accommodation and Food Services Sales                      object
Sales.Retail Sales                                               object
Employment.Firms.Total                                            int64
Employment.Firms.Women-Owned                                      int64
Employment.Firms.Men-Owned                                        int64
Employment.Firms.Minority-Owned                                  object
Employment.Firms.Nonminority-Owned                                int64
Employment.Firms.Veteran-Owned                                    int64
Employment.Firms.Nonveteran-Owned                                 int64
dtype: object

In [ ]:
#Replaces all non-numeric values with missing values for all numeric columns
words = ['Age', 'Miscellaneous', 'Housing', 'Ethnicities', 'Employment', 'Sales']
numeric_cols =  county_df[[col for col in county_df.columns if any(word in col for word in words)]]
county_df[[col for col in county_df.columns if any(word in col for word in words)]] = numeric_cols.map(lambda x: pd.to_numeric(x, errors='coerce'))

In [ ]:
county_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3139 entries, 0 to 3138
Data columns (total 44 columns):
 #   Column                                                        Non-Null Count  Dtype  
---  ------                                                        --------------  -----  
 0   County                                                        3139 non-null   object 
 1   State                                                         3139 non-null   object 
 2   mergedkeycol                                                  3139 non-null   object 
 3   Age.Percent 65 and Older                                      3139 non-null   float64
 4   Age.Percent Under 18 Years                                    3138 non-null   float64
 5   Age.Percent Under 5 Years                                     3138 non-null   float64
 6   Education.Bachelor's Degree or Higher                         3139 non-null   float64
 7   Education.High School or Higher                               3139 no

In [ ]:
def export_dataframe_to_excel(df, file_path, sheet_name='Sheet1', index=False):
    """
    Exports a Pandas DataFrame to an Excel file.

    Args:
        df (pd.DataFrame): The DataFrame to export.
        file_path (str): The path to the Excel file.
        sheet_name (str, optional): The name of the sheet. Defaults to 'Sheet1'.
        index (bool, optional): Whether to include the DataFrame index in the Excel file. Defaults to False.
    """
    try:
        df.to_excel(file_path, sheet_name=sheet_name, index=index)
        print(f"DataFrame successfully exported to {file_path}")
    except Exception as e:
        print(f"Error exporting DataFrame: {e}")
# Export the DataFrame to an Excel file
export_dataframe_to_excel(county_df, 'county_df.xlsx', sheet_name='county_demographics')

DataFrame successfully exported to county_df.xlsx


In [ ]:
#Iterate through each column for county
for col in county_df.columns:
    # Calculate the percentage of missing values
    missing_percentage = county_df[col].isna().mean()

    if missing_percentage < 0.05:
        # Drop rows if missing data is below threshold
        county_df = county_df[county_df[col].notna()]
    else:
        # Replace missing values with median if above threshold
        median_value = county_df[col].median()
        county_df.fillna({col: median_value}, inplace=True)

In [ ]:
county_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2994 entries, 0 to 3137
Data columns (total 44 columns):
 #   Column                                                        Non-Null Count  Dtype  
---  ------                                                        --------------  -----  
 0   County                                                        2994 non-null   object 
 1   State                                                         2994 non-null   object 
 2   mergedkeycol                                                  2994 non-null   object 
 3   Age.Percent 65 and Older                                      2994 non-null   float64
 4   Age.Percent Under 18 Years                                    2994 non-null   float64
 5   Age.Percent Under 5 Years                                     2994 non-null   float64
 6   Education.Bachelor's Degree or Higher                         2994 non-null   float64
 7   Education.High School or Higher                               2994 non-nul

In [ ]:
#check if there are any missing values
county_df.isna().sum()

County                                                          0
State                                                           0
mergedkeycol                                                    0
Age.Percent 65 and Older                                        0
Age.Percent Under 18 Years                                      0
Age.Percent Under 5 Years                                       0
Education.Bachelor's Degree or Higher                           0
Education.High School or Higher                                 0
Employment.Nonemployer Establishments                           0
Ethnicities.American Indian and Alaska Native Alone             0
Ethnicities.Asian Alone                                         0
Ethnicities.Black Alone                                         0
Ethnicities.Hispanic or Latino                                  0
Ethnicities.Native Hawaiian and Other Pacific Islander Alone    0
Ethnicities.Two or More Races                                   0
Ethnicities.White Alone                                         0
Ethnicities.White Alone\t not Hispanic or Latino                0
Housing.Homeownership Rate                                      0
Housing.Households                                              0
Housing.Housing Units                                           0
Housing.Median Value of Owner-Occupied Units                    0
Housing.Persons per Household                                   0
Income.Median Houseold Income                                   0
Income.Per Capita Income                                        0
Miscellaneous.Foreign Born                                      0
Miscellaneous.Land Area                                         0
Miscellaneous.Language Other than English at Home               0
Miscellaneous.Living in Same House +1 Years                     0
Miscellaneous.Manufacturers Shipments                           0
Miscellaneous.Mean Travel Time to Work                          0
Miscellaneous.Percent Female                                    0
Miscellaneous.Veterans                                          0
Population.2020 Population                                      0
Population.2010 Population                                      0
Population.Population per Square Mile                           0
Sales.Accommodation and Food Services Sales                     0
Sales.Retail Sales                                              0
Employment.Firms.Total                                          0
Employment.Firms.Women-Owned                                    0
Employment.Firms.Men-Owned                                      0
Employment.Firms.Minority-Owned                                 0
Employment.Firms.Nonminority-Owned                              0
Employment.Firms.Veteran-Owned                                  0
Employment.Firms.Nonveteran-Owned                               0
dtype: int64

In [ ]:
county_df.head()

County State       mergedkeycol  Age.Percent 65 and Older  \
0  abbevillecounty    sc  abbevillecountysc                      22.4   
1     acadiaparish    la     acadiaparishla                      15.8   
2   accomackcounty    va   accomackcountyva                      24.6   
3        adacounty    id        adacountyid                      14.9   
4      adaircounty    ia      adaircountyia                      23.0   

   Age.Percent Under 18 Years  Age.Percent Under 5 Years  \
0                        19.8                        4.7   
1                        25.8                        6.9   
2                        20.7                        5.6   
3                        23.2                        5.6   
4                        21.8                        5.6   

   Education.Bachelor's Degree or Higher  Education.High School or Higher  \
0                                   15.6                             81.7   
1                                   13.3                             79.0   
2                                   19.5                             81.5   
3                                   38.5                             95.2   
4                                   18.5                             94.2   

   Employment.Nonemployer Establishments  \
0                                 1416.0   
1                                 4533.0   
2                                 2387.0   
3                                41464.0   
4                                  609.0   

   Ethnicities.American Indian and Alaska Native Alone  ...  \
0                                                0.3    ...   
1                                                0.4    ...   
2                                                0.7    ...   
3                                                0.8    ...   
4                                                0.3    ...   

   Population.Population per Square Mile  \
0                                   51.8   
1                                   94.3   
2                                   73.8   
3                                  372.8   
4                                   13.5   

   Sales.Accommodation and Food Services Sales  Sales.Retail Sales  \
0                                      12507.0             91371.0   
1                                      52706.0            602739.0   
2                                      53568.0            348195.0   
3                                     763099.0           5766679.0   
4                                      44317.0             63002.0   

   Employment.Firms.Total  Employment.Firms.Women-Owned  \
0                    1450                           543   
1                    4664                          1516   
2                    2997                           802   
3                   41789                         14661   
4                     914                           304   

   Employment.Firms.Men-Owned  Employment.Firms.Minority-Owned  \
0                         689                            317.0   
1                        2629                            705.0   
2                        1716                            335.0   
3                       19409                           3099.0   
4                         499                              0.0   

   Employment.Firms.Nonminority-Owned  Employment.Firms.Veteran-Owned  \
0                                1080                             187   
1                                3734                             388   
2                                2560                             212   
3                               36701                            3803   
4                                 861                             185   

   Employment.Firms.Nonveteran-Owned  
0                               1211  
1                               4007  
2                               2536  
3                              35132  
4                                

##Cleaning Food_df

###Checking DataTypes

In [ ]:
food_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3142 entries, 0 to 3141
Data columns (total 26 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   County                                         3142 non-null   object 
 1   Population                                     3142 non-null   int64  
 2   State                                          3142 non-null   object 
 3   mergedkeycol                                   3142 non-null   object 
 4   Housing Data.Residing in Group Quarters        3142 non-null   int64  
 5   Housing Data.Total Housing Units               3142 non-null   int64  
 6   Vehicle Access.1 Mile                          3142 non-null   int64  
 7   Vehicle Access.1/2 Mile                        3142 non-null   int64  
 8   Vehicle Access.10 Miles                        2807 non-null   float64
 9   Vehicle Access.20 Miles                        3142 

###Checking for Null Values

In [ ]:
for col in food_df.columns[4:26]:
    print(f"'{col}': {food_df[col].isnull().sum()}")

'Housing Data.Residing in Group Quarters': 0
'Housing Data.Total Housing Units': 0
'Vehicle Access.1 Mile': 0
'Vehicle Access.1/2 Mile': 0
'Vehicle Access.10 Miles': 335
'Vehicle Access.20 Miles': 0
'Low Access Numbers.Children.1 Mile': 0
'Low Access Numbers.Children.1/2 Mile': 0
'Low Access Numbers.Children.10 Miles': 0
'Low Access Numbers.Children.20 Miles': 0
'Low Access Numbers.Low Income People.1 Mile': 0
'Low Access Numbers.Low Income People.1/2 Mile': 0
'Low Access Numbers.Low Income People.10 Miles': 0
'Low Access Numbers.Low Income People.20 Miles': 0
'Low Access Numbers.People.1 Mile': 156
'Low Access Numbers.People.1/2 Mile': 0
'Low Access Numbers.People.10 Miles': 0
'Low Access Numbers.People.20 Miles': 0
'Low Access Numbers.Seniors.1 Mile': 0
'Low Access Numbers.Seniors.1/2 Mile': 658
'Low Access Numbers.Seniors.10 Miles': 0
'Low Access Numbers.Seniors.20 Miles': 0


In [ ]:
pd.isna(food_df['Vehicle Access.10 Miles'][19])

True

In [ ]:
pd.isna(food_df['Low Access Numbers.People.1 Mile'][2])

True

In [ ]:
for col in food_df.columns[4:26]:
    # Count the occurrences of NaN values in the column
    nan_count = food_df[col].isnull().sum()

    # Print the column name and the count of NaN values
    print(f"'{col}': {nan_count}")

'Housing Data.Residing in Group Quarters': 0
'Housing Data.Total Housing Units': 0
'Vehicle Access.1 Mile': 0
'Vehicle Access.1/2 Mile': 0
'Vehicle Access.10 Miles': 335
'Vehicle Access.20 Miles': 0
'Low Access Numbers.Children.1 Mile': 0
'Low Access Numbers.Children.1/2 Mile': 0
'Low Access Numbers.Children.10 Miles': 0
'Low Access Numbers.Children.20 Miles': 0
'Low Access Numbers.Low Income People.1 Mile': 0
'Low Access Numbers.Low Income People.1/2 Mile': 0
'Low Access Numbers.Low Income People.10 Miles': 0
'Low Access Numbers.Low Income People.20 Miles': 0
'Low Access Numbers.People.1 Mile': 156
'Low Access Numbers.People.1/2 Mile': 0
'Low Access Numbers.People.10 Miles': 0
'Low Access Numbers.People.20 Miles': 0
'Low Access Numbers.Seniors.1 Mile': 0
'Low Access Numbers.Seniors.1/2 Mile': 658
'Low Access Numbers.Seniors.10 Miles': 0
'Low Access Numbers.Seniors.20 Miles': 0


###TRIAL READCSV ISSUE

In [ ]:
mode_food_df = pd.read_csv("https://raw.githubusercontent.com/nidhicodes4045/datascience442/refs/heads/main/Food_access_mod.csv", dtype=str, keep_default_na=False)

In [ ]:
from IPython.display import display

# Checking different types of null representations
null_counts = {
    "Explicit NaN (pd.isna)": mode_food_df.applymap(lambda x: pd.isna(x)).sum(),
    "Empty Strings ('')": (mode_food_df == "").sum(),
    "Whitespace Strings": (mode_food_df.applymap(lambda x: x.strip() if isinstance(x, str) else x) == "").sum(),
    "Explicit 'N/A' Strings": (mode_food_df.applymap(lambda x: x.lower() if isinstance(x, str) else x) == "n/a").sum(),
    "Explicit 'NULL' Strings": (mode_food_df.applymap(lambda x: x.lower() if isinstance(x, str) else x) == "null").sum()
}

# Combine results into a DataFrame
null_summary = pd.DataFrame(null_counts)
# Show the DataFrame
display(null_summary)


<ipython-input-485-9f05aeb7ef44>:5: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  "Explicit NaN (pd.isna)": mode_food_df.applymap(lambda x: pd.isna(x)).sum(),
<ipython-input-485-9f05aeb7ef44>:7: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  "Whitespace Strings": (mode_food_df.applymap(lambda x: x.strip() if isinstance(x, str) else x) == "").sum(),
<ipython-input-485-9f05aeb7ef44>:8: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  "Explicit 'N/A' Strings": (mode_food_df.applymap(lambda x: x.lower() if isinstance(x, str) else x) == "n/a").sum(),
<ipython-input-485-9f05aeb7ef44>:9: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  "Explicit 'NULL' Strings": (mode_food_df.applymap(lambda x: x.lower() if isinstance(x, str) else x) == "null").sum()


Explicit NaN (pd.isna)  \
County                                                              0   
Population                                                          0   
State                                                               0   
Housing Data.Residing in Group Quarters                             0   
Housing Data.Total Housing Units                                    0   
Vehicle Access.1 Mile                                               0   
Vehicle Access.1/2 Mile                                             0   
Vehicle Access.10 Miles                                             0   
Vehicle Access.20 Miles                                             0   
Low Access Numbers.Children.1 Mile                                  0   
Low Access Numbers.Children.1/2 Mile                                0   
Low Access Numbers.Children.10 Miles                                0   
Low Access Numbers.Children.20 Miles                                0   
Low Access Numbers.Low Income People.1 Mile                         0   
Low Access Numbers.Low Income People.1/2 Mile                       0   
Low Access Numbers.Low Income People.10 Miles                       0   
Low Access Numbers.Low Income People.20 Miles                       0   
Low Access Numbers.People.1 Mile                                    0   
Low Access Numbers.People.1/2 Mile                                  0   
Low Access Numbers.People.10 Miles                                  0   
Low Access Numbers.People.20 Miles                                  0   
Low Access Numbers.Seniors.1 Mile                                   0   
Low Access Numbers.Seniors.1/2 Mile                                 0   
Low Access Numbers.Seniors.10 Miles                                 0   
Low Access Numbers.Seniors.20 Miles                                 0   

                                               Empty Strings ('')  \
County                                                          0   
Population                                                      0   
State                                                           0   
Housing Data.Residing in Group Quarters                         0   
Housing Data.Total Housing Units                                0   
Vehicle Access.1 Mile                                           0   
Vehicle Access.1/2 Mile                                         0   
Vehicle Access.10 Miles                                       335   
Vehicle Access.20 Miles                                         0   
Low Access Numbers.Children.1 Mile                              0   
Low Access Numbers.Children.1/2 Mile                            0   
Low Access Numbers.Children.10 Miles                            0   
Low Access Numbers.Children.20 Miles                            0   
Low Access Numbers.Low Income People.1 Mile                     0   
Low Access Numbers.Low Income People.1/2 Mile                   0   
Low Access Numbers.Low Income People.10 Miles                   0   
Low Access Numbers.Low Income People.20 Miles                   0   
Low Access Numbers.People.1 Mile                                0   
Low Access Numbers.People.1/2 Mile                              0   
Low Access Numbers.People.10 Miles                              0   
Low Access Numbers.People.20 Miles                              0   
Low Access Numbers.Seniors.1 Mile                               0   
Low Access Numbers.Seniors.1/2 Mile                             0   
Low Access Numbers.Seniors.10 Miles                             0   
Low Access Numbers.Seniors.20 Miles                             0   

                                               Whitespace Strings  \
County                                                          0   
Population                                                      0   
State                                                           0   
Housing Data.Residing in Group Quarters                         0   
Housing 

In [ ]:
mode_food_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3142 entries, 0 to 3141
Data columns (total 25 columns):
 #   Column                                         Non-Null Count  Dtype 
---  ------                                         --------------  ----- 
 0   County                                         3142 non-null   object
 1   Population                                     3142 non-null   object
 2   State                                          3142 non-null   object
 3   Housing Data.Residing in Group Quarters        3142 non-null   object
 4   Housing Data.Total Housing Units               3142 non-null   object
 5   Vehicle Access.1 Mile                          3142 non-null   object
 6   Vehicle Access.1/2 Mile                        3142 non-null   object
 7   Vehicle Access.10 Miles                        3142 non-null   object
 8   Vehicle Access.20 Miles                        3142 non-null   object
 9   Low Access Numbers.Children.1 Mile             3142 non-null   

In [ ]:
# Convert columns 4-25 (except 8, 18, and 23) to appropriate numerical types
cols_to_convert = [mode_food_df.columns[i] for i in range(3, 25) if i not in [7, 17, 22]]

# Convert to float since some numeric values may have decimals
mode_food_df[cols_to_convert] = mode_food_df[cols_to_convert].apply(pd.to_numeric, errors='coerce')

# Display the updated data types
mode_food_df.dtypes


County                                           object
Population                                       object
State                                            object
Housing Data.Residing in Group Quarters           int64
Housing Data.Total Housing Units                  int64
Vehicle Access.1 Mile                             int64
Vehicle Access.1/2 Mile                           int64
Vehicle Access.10 Miles                          object
Vehicle Access.20 Miles                           int64
Low Access Numbers.Children.1 Mile                int64
Low Access Numbers.Children.1/2 Mile              int64
Low Access Numbers.Children.10 Miles              int64
Low Access Numbers.Children.20 Miles              int64
Low Access Numbers.Low Income People.1 Mile       int64
Low Access Numbers.Low Income People.1/2 Mile     int64
Low Access Numbers.Low Income People.10 Miles     int64
Low Access Numbers.Low Income People.20 Miles     int64
Low Access Numbers.People.1 Mile                 object
Low Access Numbers.People.1/2 Mile                int64
Low Access Numbers.People.10 Miles                int64
Low Access Numbers.People.20 Miles                int64
Low Access Numbers.Seniors.1 Mile                 int64
Low Access Numbers.Seniors.1/2 Mile              object
Low Access Numbers.Seniors.10 Miles               int64
Low Access Numbers.Seniors.20 Miles               int64
dtype: object

## Summarize Missing value columns with temp df

In [ ]:
df_temp = mode_food_df.copy()
missing_cols = [mode_food_df.columns[7], mode_food_df.columns[17], mode_food_df.columns[22]]
df_temp[missing_cols] = df_temp[missing_cols].replace(["", " ", "N/A", "n/a", "NULL", "null"], None)
df_temp[missing_cols] = df_temp[missing_cols].apply(pd.to_numeric, errors='coerce').fillna(0)
summary_stats = df_temp[missing_cols].describe()
from IPython.display import display
display(summary_stats)

Vehicle Access.10 Miles  Low Access Numbers.People.1 Mile  \
count              3142.000000                       3142.000000   
mean                 27.729472                      37205.573838   
std                  78.577763                      64074.586653   
min                   0.000000                          0.000000   
25%                   0.000000                       6282.500000   
50%                   7.000000                      16444.500000   
75%                  30.000000                      39954.500000   
max                1826.000000                     903299.000000   

       Low Access Numbers.Seniors.1/2 Mile  
count                          3142.000000  
mean                           7162.901018  
std                           17148.344674  
min                               0.000000  
25%                             408.250000  
50%                            2416.000000  
75%                            6152.250000  
max                          319028.000000

#Analysis of the Summary

##Column: Vehicle Access.10 Miles (Column 7)

- Mean: 27.73
- Median (50%): 7.0
- Min: 0, Max: 1826
- Skewed distribution (high max value compared to median)
- Best Replacement: Median (7.0) (since mean is influenced by extreme values)

##Column: Low Access Numbers.People.1 Mile (Column 17)

- Mean: 37,205
- Median (50%): 16,444
- Min: 0, Max: 903,299
- Highly skewed (huge gap between min, median, and max)
- Best Replacement: Median (16,444.5) (again, to avoid extreme values pulling the mean)

##Column: Low Access Numbers.Seniors.1/2 Mile (Column 22)

- Mean: 7,162
- Median (50%): 2,416
- Min: 0, Max: 319,028
- Highly skewed (huge max value)
- Best Replacement: Median (2,416.0)

#Final Replacement Approach

- Replace missing values in Vehicle Access.10 Miles with 7.0
- Replace missing values in Low Access Numbers.People.1 Mile with 16,444.5
- Replace missing values in Low Access Numbers.Seniors.1/2 Mile with 2,416.

In [ ]:
# Define replacement values based on median
replacement_values = {
    mode_food_df.columns[7]: 7.0,          # Vehicle Access.10 Miles
    mode_food_df.columns[17]: 16444.5,     # Low Access Numbers.People.1 Mile
    mode_food_df.columns[22]: 2416.0       # Low Access Numbers.Seniors.1/2 Mile
}
columns_to_replace = list(replacement_values.keys())

# Normalize missing values
mode_food_df[columns_to_replace] = mode_food_df[columns_to_replace].replace(["", " ", "N/A", "n/a", "NULL", "null"], None)

# Convert to numeric and fill missing values
mode_food_df[columns_to_replace] = mode_food_df[columns_to_replace].apply(pd.to_numeric, errors='coerce').fillna(replacement_values)

# Verify the changes
from IPython.display import display
display(mode_food_df[replacement_values.keys()].describe())

Vehicle Access.10 Miles  Low Access Numbers.People.1 Mile  \
count              3142.000000                       3142.000000   
mean                 28.475812                      38022.041693   
std                  78.343661                      63698.844970   
min                   0.000000                          0.000000   
25%                   2.000000                       7762.750000   
50%                   7.000000                      16445.750000   
75%                  30.000000                      39954.500000   
max                1826.000000                     903299.000000   

       Low Access Numbers.Seniors.1/2 Mile  
count                          3142.000000  
mean                           7668.861553  
std                           16964.133907  
min                              12.000000  
25%                            2042.000000  
50%                            2417.000000  
75%                            6152.250000  
max                          319028.000000

In [ ]:
mode_food_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3142 entries, 0 to 3141
Data columns (total 25 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   County                                         3142 non-null   object 
 1   Population                                     3142 non-null   object 
 2   State                                          3142 non-null   object 
 3   Housing Data.Residing in Group Quarters        3142 non-null   int64  
 4   Housing Data.Total Housing Units               3142 non-null   int64  
 5   Vehicle Access.1 Mile                          3142 non-null   int64  
 6   Vehicle Access.1/2 Mile                        3142 non-null   int64  
 7   Vehicle Access.10 Miles                        3142 non-null   float64
 8   Vehicle Access.20 Miles                        3142 non-null   int64  
 9   Low Access Numbers.Children.1 Mile             3142 

In [ ]:
mode_food_df['Low Access Numbers.People.1 Mile'][2]

16444.5

In [ ]:
mode_food_df['Vehicle Access.10 Miles'][19]

7.0

In [ ]:
mode_food_df['Low Access Numbers.Seniors.1/2 Mile'][3139]

2416.0

In [ ]:
mode_food_df.head()

County Population    State  \
0  Autauga County      54571  Alabama   
1  Baldwin County     182265  Alabama   
2  Barbour County      27457  Alabama   
3     Bibb County      22915  Alabama   
4   Blount County      57322  Alabama   

   Housing Data.Residing in Group Quarters  Housing Data.Total Housing Units  \
0                                      455                             20221   
1                                     2307                             73180   
2                                     3193                              9820   
3                                     2224                              7953   
4                                      489                             21578   

   Vehicle Access.1 Mile  Vehicle Access.1/2 Mile  Vehicle Access.10 Miles  \
0                    834                     1045                    222.0   
1                   1653                     2178                     32.0   
2                    545                      742                    201.0   
3                    312                      441                      0.0   
4                    752                      822                      0.0   

   Vehicle Access.20 Miles  Low Access Numbers.Children.1 Mile  ...  \
0                        0                                9973  ...   
1                        0                               30633  ...   
2                        0                                3701  ...   
3                        0                                4198  ...   
4                        0                               12575  ...   

   Low Access Numbers.Low Income People.10 Miles  \
0                                           2307   
1                                            846   
2                                           2440   
3                                            102   
4                                              0   

   Low Access Numbers.Low Income People.20 Miles  \
0                                              0   
1                                              0   
2                                              0   
3                                              0   
4                                              0   

   Low Access Numbers.People.1 Mile  Low Access Numbers.People.1/2 Mile  \
0                           37424.0                               49497   
1                          132442.0                              165616   
2                           16444.5                               23762   
3                           17560.0                               20989   
4                           50848.0                               54933   

   Low Access Numbers.People.10 Miles  Low Access Numbers.People.20 Miles  \
0                                5119                                   0   
1                                2308                                   0   
2                                4643                                   0   
3                                 365                                   0   
4                                   0                                   0   

   Low Access Numbers.Seniors.1 Mile  Low Access Numbers.Seniors.1/2 Mile  \
0                               4393                               5935.0   
1                              21828                              27241.0   
2                               2537                               3348.0   
3                               2262                               2630.0   
4                               7114                               7810.0   

   Low Access Numbers.Seniors.10 Miles  Low Access Numbers.Seniors.20 Miles  
0                                  707                                    0  
1                                  390                                    0  
2                                  629                                    0  
3                                   72                                  

#Fixing primary key for the cleaned food dataframe (mode_food_df)

In [ ]:
#Apply the lowercase and strip functions to the State and County cols county df
#Apply the preprocessing functions to food df
mode_food_df["County"] = mode_food_df["County"].apply(lowerandstriptext)
mode_food_df["State"] = mode_food_df["State"].apply(convertstate)
mode_food_df.head()

County Population State  Housing Data.Residing in Group Quarters  \
0  autaugacounty      54571    al                                      455   
1  baldwincounty     182265    al                                     2307   
2  barbourcounty      27457    al                                     3193   
3     bibbcounty      22915    al                                     2224   
4   blountcounty      57322    al                                      489   

   Housing Data.Total Housing Units  Vehicle Access.1 Mile  \
0                             20221                    834   
1                             73180                   1653   
2                              9820                    545   
3                              7953                    312   
4                             21578                    752   

   Vehicle Access.1/2 Mile  Vehicle Access.10 Miles  Vehicle Access.20 Miles  \
0                     1045                    222.0                        0   
1                     2178                     32.0                        0   
2                      742                    201.0                        0   
3                      441                      0.0                        0   
4                      822                      0.0                        0   

   Low Access Numbers.Children.1 Mile  ...  \
0                                9973  ...   
1                               30633  ...   
2                                3701  ...   
3                                4198  ...   
4                               12575  ...   

   Low Access Numbers.Low Income People.10 Miles  \
0                                           2307   
1                                            846   
2                                           2440   
3                                            102   
4                                              0   

   Low Access Numbers.Low Income People.20 Miles  \
0                                              0   
1                                              0   
2                                              0   
3                                              0   
4                                              0   

   Low Access Numbers.People.1 Mile  Low Access Numbers.People.1/2 Mile  \
0                           37424.0                               49497   
1                          132442.0                              165616   
2                           16444.5                               23762   
3                           17560.0                               20989   
4                           50848.0                               54933   

   Low Access Numbers.People.10 Miles  Low Access Numbers.People.20 Miles  \
0                                5119                                   0   
1                                2308                                   0   
2                                4643                                   0   
3                                 365                                   0   
4                                   0                                   0   

   Low Access Numbers.Seniors.1 Mile  Low Access Numbers.Seniors.1/2 Mile  \
0                               4393                               5935.0   
1                              21828                              27241.0   
2                               2537                               3348.0   
3                               2262                               2630.0   
4                               7114                               7810.0   

   Low Access Numbers.Seniors.10 Miles  Low Access Numbers.Seniors.20 Miles  
0                                  707                                    0  
1                                  390                                    0  
2                                  629                                    0  
3                                   72                                    0  
4                

In [ ]:
#Verify if any 'Nones' exist before merging
check_state_mode_food_df = (mode_food_df['State'] == 'None').sum()
print("The number of 'None' in mode_food_df: " + str(check_state_mode_food_df))

The number of 'None' in mode_food_df: 0


In [ ]:
#Insert merged columns at index 2 and 3 for county and food dfs respectively
mode_food_df.insert(3, "mergedkeycol", mode_food_df['County'] + mode_food_df['State'])

In [ ]:
mode_food_df.head()

County Population State     mergedkeycol  \
0  autaugacounty      54571    al  autaugacountyal   
1  baldwincounty     182265    al  baldwincountyal   
2  barbourcounty      27457    al  barbourcountyal   
3     bibbcounty      22915    al     bibbcountyal   
4   blountcounty      57322    al   blountcountyal   

   Housing Data.Residing in Group Quarters  Housing Data.Total Housing Units  \
0                                      455                             20221   
1                                     2307                             73180   
2                                     3193                              9820   
3                                     2224                              7953   
4                                      489                             21578   

   Vehicle Access.1 Mile  Vehicle Access.1/2 Mile  Vehicle Access.10 Miles  \
0                    834                     1045                    222.0   
1                   1653                     2178                     32.0   
2                    545                      742                    201.0   
3                    312                      441                      0.0   
4                    752                      822                      0.0   

   Vehicle Access.20 Miles  ...  \
0                        0  ...   
1                        0  ...   
2                        0  ...   
3                        0  ...   
4                        0  ...   

   Low Access Numbers.Low Income People.10 Miles  \
0                                           2307   
1                                            846   
2                                           2440   
3                                            102   
4                                              0   

   Low Access Numbers.Low Income People.20 Miles  \
0                                              0   
1                                              0   
2                                              0   
3                                              0   
4                                              0   

   Low Access Numbers.People.1 Mile  Low Access Numbers.People.1/2 Mile  \
0                           37424.0                               49497   
1                          132442.0                              165616   
2                           16444.5                               23762   
3                           17560.0                               20989   
4                           50848.0                               54933   

   Low Access Numbers.People.10 Miles  Low Access Numbers.People.20 Miles  \
0                                5119                                   0   
1                                2308                                   0   
2                                4643                                   0   
3                                 365                                   0   
4                                   0                                   0   

   Low Access Numbers.Seniors.1 Mile  Low Access Numbers.Seniors.1/2 Mile  \
0                               4393                               5935.0   
1                              21828                              27241.0   
2                               2537                               3348.0   
3                               2262                               2630.0   
4                               7114                               7810.0   

   Low Access Numbers.Seniors.10 Miles  Low Access Numbers.Seniors.20 Miles  
0                                  707                                    0  
1                                  390                                    0  
2                                  629                                    0  
3                                   72                                    0  
4                                    0                                    0  

[5 rows x 26 columns]

In [ ]:
county_df.head()

County State       mergedkeycol  Age.Percent 65 and Older  \
0  abbevillecounty    sc  abbevillecountysc                      22.4   
1     acadiaparish    la     acadiaparishla                      15.8   
2   accomackcounty    va   accomackcountyva                      24.6   
3        adacounty    id        adacountyid                      14.9   
4      adaircounty    ia      adaircountyia                      23.0   

   Age.Percent Under 18 Years  Age.Percent Under 5 Years  \
0                        19.8                        4.7   
1                        25.8                        6.9   
2                        20.7                        5.6   
3                        23.2                        5.6   
4                        21.8                        5.6   

   Education.Bachelor's Degree or Higher  Education.High School or Higher  \
0                                   15.6                             81.7   
1                                   13.3                             79.0   
2                                   19.5                             81.5   
3                                   38.5                             95.2   
4                                   18.5                             94.2   

   Employment.Nonemployer Establishments  \
0                                 1416.0   
1                                 4533.0   
2                                 2387.0   
3                                41464.0   
4                                  609.0   

   Ethnicities.American Indian and Alaska Native Alone  ...  \
0                                                0.3    ...   
1                                                0.4    ...   
2                                                0.7    ...   
3                                                0.8    ...   
4                                                0.3    ...   

   Population.Population per Square Mile  \
0                                   51.8   
1                                   94.3   
2                                   73.8   
3                                  372.8   
4                                   13.5   

   Sales.Accommodation and Food Services Sales  Sales.Retail Sales  \
0                                      12507.0             91371.0   
1                                      52706.0            602739.0   
2                                      53568.0            348195.0   
3                                     763099.0           5766679.0   
4                                      44317.0             63002.0   

   Employment.Firms.Total  Employment.Firms.Women-Owned  \
0                    1450                           543   
1                    4664                          1516   
2                    2997                           802   
3                   41789                         14661   
4                     914                           304   

   Employment.Firms.Men-Owned  Employment.Firms.Minority-Owned  \
0                         689                            317.0   
1                        2629                            705.0   
2                        1716                            335.0   
3                       19409                           3099.0   
4                         499                              0.0   

   Employment.Firms.Nonminority-Owned  Employment.Firms.Veteran-Owned  \
0                                1080                             187   
1                                3734                             388   
2                                2560                             212   
3                               36701                            3803   
4                                 861                             185   

   Employment.Firms.Nonveteran-Owned  
0                               1211  
1                               4007  
2                               2536  
3                              35132  
4                                

In [ ]:
mode_food_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3142 entries, 0 to 3141
Data columns (total 26 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   County                                         3142 non-null   object 
 1   Population                                     3142 non-null   object 
 2   State                                          3142 non-null   object 
 3   mergedkeycol                                   3142 non-null   object 
 4   Housing Data.Residing in Group Quarters        3142 non-null   int64  
 5   Housing Data.Total Housing Units               3142 non-null   int64  
 6   Vehicle Access.1 Mile                          3142 non-null   int64  
 7   Vehicle Access.1/2 Mile                        3142 non-null   int64  
 8   Vehicle Access.10 Miles                        3142 non-null   float64
 9   Vehicle Access.20 Miles                        3142 

In [ ]:
county_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2994 entries, 0 to 3137
Data columns (total 44 columns):
 #   Column                                                        Non-Null Count  Dtype  
---  ------                                                        --------------  -----  
 0   County                                                        2994 non-null   object 
 1   State                                                         2994 non-null   object 
 2   mergedkeycol                                                  2994 non-null   object 
 3   Age.Percent 65 and Older                                      2994 non-null   float64
 4   Age.Percent Under 18 Years                                    2994 non-null   float64
 5   Age.Percent Under 5 Years                                     2994 non-null   float64
 6   Education.Bachelor's Degree or Higher                         2994 non-null   float64
 7   Education.High School or Higher                               2994 non-nul

#Data Merging
Find the common keys in the two dataframes and put it into a list

In [ ]:

common_keys = list(set(mode_food_df['mergedkeycol']) & set(county_df['mergedkeycol'])) # & means intersection
print(len(common_keys))
print("Common keys between the two data frames: " + str(common_keys))

# don't want duplicate county and state columns after merging
county_df_cols_to_keep = [col for col in county_df.columns if col not in ['County', 'State']]
county_df_to_merge = county_df[county_df_cols_to_keep]

merged_df = pd.merge(mode_food_df, county_df_to_merge, on='mergedkeycol', how='inner')
merged_df

2993
Common keys between the two data frames: ['leoncountytx', 'goliadcountytx', 'warrencountyga', 'adaircountyky', 'chickasawcountyia', 'bristolcountyri', 'barbercountyks', 'santabarbaracountyca', 'lafayettecountyfl', 'mccormickcountysc', 'eastlandcountytx', 'gregorycountysd', 'perrycountyil', 'lambcountytx', 'shastacountyca', 'boundarycountyid', 'senecacountyoh', 'bradleycountyar', 'schleycountyga', 'leflorecountyms', 'montcalmcountymi', 'houstoncountyga', 'ironcountymi', 'claiborneparishla', 'hopkinscountytx', 'whitecountyar', 'scottcountyks', 'whitecountyil', 'limestonecountytx', 'sanmateocountyca', 'texascountyok', 'colbertcountyal', 'gentrycountymo', 'baileycountytx', 'laurelcountyky', 'washingtoncountypa', 'whitleycountyky', 'dutchesscountyny', 'salinecountyar', 'hancockcountyga', 'daviecountync', 'okfuskeecountyok', 'hendrycountyfl', 'falloncountymt', 'rockdalecountyga', 'pulaskicountyar', 'roanecountywv', 'bryancountyok', 'gwinnettcountyga', 'madisoncountyky', 'laplatacountyco

County Population State        mergedkeycol  \
0        autaugacounty      54571    al     autaugacountyal   
1        baldwincounty     182265    al     baldwincountyal   
2        barbourcounty      27457    al     barbourcountyal   
3           bibbcounty      22915    al        bibbcountyal   
4         blountcounty      57322    al      blountcountyal   
...                ...        ...   ...                 ...   
2988  sweetwatercounty      43806    wy  sweetwatercountywy   
2989       tetoncounty      21294    wy       tetoncountywy   
2990       uintacounty      21118    wy       uintacountywy   
2991    washakiecounty       8533    wy    washakiecountywy   
2992      westoncounty       7208    wy      westoncountywy   

      Housing Data.Residing in Group Quarters  \
0                                         455   
1                                        2307   
2                                        3193   
3                                        2224   
4                                         489   
...                                       ...   
2988                                      679   
2989                                      271   
2990                                      270   
2991                                      140   
2992                                      313   

      Housing Data.Total Housing Units  Vehicle Access.1 Mile  \
0                                20221                    834   
1                                73180                   1653   
2                                 9820                    545   
3                                 7953                    312   
4                                21578                    752   
...                                ...                    ...   
2988                             16475                    284   
2989                              8973                     74   
2990                              7668                    175   
2991                              3492                     37   
2992                              3021                     22   

      Vehicle Access.1/2 Mile  Vehicle Access.10 Miles  \
0                        1045                    222.0   
1                        2178                     32.0   
2                         742                    201.0   
3                         441                      0.0   
4                         822                      0.0   
...                       ...                      ...   
2988                      372                     18.0   
2989                      144                      0.0   
2990                      283                      9.0   
2991                      106                      7.0   
2992                       43                      2.0   

      Vehicle Access.20 Miles  ...  Population.Population per Square Mile  \
0                           0  ...                                   91.8   
1                           0  ...                                  114.6   
2                           0  ...                                   31.0   
3                           0  ...                                   36.8   
4                           0  ...                                   88.9   
...                       ...  ...                                    ...   
2988                       13  ...                                    4.2   
2989                        0  ...                                    5.3   
2990                        0  ...                                   10.1   
2991                        6  ...                                    3.8   
2992                        0  ...                                    3.0   

      Sales.Accommodation and Food Services Sales  Sales.Retail Sales  \
0                                         93431.0            607854.0   
1                                        560602.0           3145789.0   
2                                         26157.

In [ ]:
merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2993 entries, 0 to 2992
Data columns (total 67 columns):
 #   Column                                                        Non-Null Count  Dtype  
---  ------                                                        --------------  -----  
 0   County                                                        2993 non-null   object 
 1   Population                                                    2993 non-null   object 
 2   State                                                         2993 non-null   object 
 3   mergedkeycol                                                  2993 non-null   object 
 4   Housing Data.Residing in Group Quarters                       2993 non-null   int64  
 5   Housing Data.Total Housing Units                              2993 non-null   int64  
 6   Vehicle Access.1 Mile                                         2993 non-null   int64  
 7   Vehicle Access.1/2 Mile                                       2993 no